design/one-pager-merge-jets-proposal.md

```diff
@@ -0,0 +1,428 @@
+them are maintained by their cloud vendor. The aim of this document is to
+capture a decision about the future of these provider implementations with the
+goals stated above in mind.
+
+## Proposal
+
+There were three options in the [provider strategy doc][strategy-doc]:
+* Option A: Completely separate providers.
+  * The initial decision.
+* Option B: One provider but Jet-based resources in their own API group.
+* Option C: One provider and no Jet vs classic difference. Only API version
+  would be bumped if the implementation changes.
+
+We should go with the Option C, i.e. use Terrajet in
+[`provider-aws`][provider-aws], [`provider-azure`][provider-azure] and
+[`provider-gcp`][provider-gcp] as another way of generating CRDs and controllers
+so that they can reach the goal of coverage really fast. Then the community can
+keep investing in that single provider and make each CRD mature enough to depend
+on in production. In this way, we can increase the usage by increasing both the
+coverage and the maturity of the CRDs in a consolidated way. This increase in
+usage can attract cloud vendors to come and join the maintainership of these
+providers. Once they do, they can either progressively convert the
+implementation of the controllers to their choice of technology or start with
+new providers which we can suggest people to use and we can provide migration
+tooling from the community-maintained one. In either case, the users would see a
+clear path forward and we'd not waste any efforts by the community.
+
+Note that we are making two decisions here:
+* There will be a single provider.
+* The single provider will be the classic one and we'll add Terrajet CRDs on
+  top.
+
+The order of execution will be roughly as following, taking
+[`provider-aws`][provider-aws] as an example:
+1. Add Terrajet pipeline to the provider repository.
+1. Copy the manual resource configurations from provider-jet-aws and introduce
+  the CRDs of those configurations in their new API group, without changing
+  their version, i.e. `v1alpha2`.
+1. Generate the full set of CRDs in `release-x.x-preview` branch similar to jet
+  repository and include those non-configured ones only in `-preview` suffixed
+  version.
+1. Write up a guide or tool to help the current users of non-preview version of
+  Jet providers to migrate to [`provider-aws`][provider-aws].
+
+We will also update the guidelines about how to add a resource roughly with the
+following:
+* If you are adding a new resource, use Terrajet pipeline unless there is a
+  blocker.
+* If you encounter a bug that cannot be solved with Terrajet in a meaningful
+  timeframe, consider converting its controller to classic hand-written
+  controller and bump the API version.
+* If an existing classic CRD (hand-written or ACK-generated) has a bug, you
+  should consider converting its controller to Terrajet first.
+* Existing classic CRDs that are on alpha version will be converted to Terrajet
```

Comment:
I don't quite understand this. If we have a classic CRD that has been well tested by users and we are thinking in bumping it to beta (when do we should bump to beta?), we should convert it to terrajet? In my opinion, that would "throw away" the well tested part in the alpha version. Looks like we are creating a jet resource in beta instead of alpha. Am I missing something?

Reply:
> In my opinion, that would "throw away" the well tested part of the alpha version.

I'm on the fence about this but yes, this is one of the implications of aggressiveness in converging on Terrajet, which is discussed in [this thread](https://github.com/crossplane/crossplane/pull/2930#discussion_r838672541). Even beta ones are suggested to convert after extensive testing. The initial proposal was along the lines of `do not convert to Terrajet unless there is a bug` but now it's more like trying everything we can to push to Terrajet. This one is more on the technical enforcement side.  Do you find this too aggressive?

> Looks like we are creating a jet resource in beta instead of alpha.

I'd imagine something like `v1alpha1 (classic)` -> `v1alpha2 (terrajet)` -> and continue with terrajet.

Reply:
Oh great 😄  Yes it definitely involves less effort to start and also maintain since TF providers are very mature but the conversion does take some effort due to behavior tests. Please feel free to approve or decline the PR as you are a maintainer in one of the affected repositories 🙂